File: src/router/node.js
/**
 * Created by igor on 29.03.17.
 */

"use strict";

const BaseNode = require('./baseNode');

class Node extends BaseNode {
    constructor (parent) {
        super(parent);
        this.children = [];
        this.position = 0;
    }

    get length () {
        return this.children.length;
    }

    add (node) {
        this.children.push(node)
    }

    first () {
        this.position = 0;
    }

    isLast () {
        return this.children.length === (this.position + 1);
    }

    next () {
        return this.children[this.position++];
    }
}

module.exports = Node;

File: src/pkg/router/node.go
/**
 * Created by I. Navrotskyj on 19.08.17.
 */

package router


type Node struct {
	BaseNode
	position int
	children []App
}

func (n *Node) getLength() int  {
	return len(n.children)
}

func (n *Node) Add (newNode App)  {
	n.children = append(n.children, newNode)
}

func (n *Node) setFirst ()  {
	n.position = 0
}

func (n *Node) Next() (newNode App)  {
	if n.position < n.getLength() {
		newNode = n.children[n.position]
		n.position++
	}
	return
}

func NewNode(parent *Node) *Node {
	n := &Node{}
	n.setParentNode(parent)

	return n
}